src/main/java/uk/gov/companieshouse/api/accounts/service/impl/CompanyAccountServiceImpl.java

```diff
@@ -1,33 +1,98 @@
 import uk.gov.companieshouse.api.accounts.transformer.GenericTransformer;
 
 @Service
 public class CompanyAccountServiceImpl extends
         AbstractServiceImpl<CompanyAccount, CompanyAccountEntity> implements CompanyAccountService {
 
+    @Autowired
+    private TransactionManager transactionManager;
+
     @Autowired
     public CompanyAccountServiceImpl(
             @Qualifier("companyAccountRepository") MongoRepository<CompanyAccountEntity, String> mongoRepository,
             @Qualifier("companyAccountTransformer") GenericTransformer<CompanyAccount, CompanyAccountEntity> transformer) {
         super(mongoRepository, transformer);
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    public CompanyAccount createCompanyAccount(CompanyAccount companyAccount) {
+        HttpServletRequest request = getRequestFromContext();
+        ResponseEntity<Transaction> transaction = getTransactionFromSession(request);
+
+        addKind(companyAccount);
+        addEtag(companyAccount);
+
+        CompanyAccountEntity companyAccountEntity = genericTransformer.transform(companyAccount);
+
+        addEntityID(companyAccountEntity);
+        addEntityLinks(companyAccountEntity, transaction);
+
+        mongoRepository.insert(companyAccountEntity);
+
+        transactionManager.updateTransaction(transaction.getBody().getId(), request.getHeader("X-Request-Id"), getCompanyAccountSelfLink(companyAccountEntity));
+
+        return genericTransformer.transform(companyAccountEntity);
+    }
+
+    public void addEntityLinks(CompanyAccountEntity entity, ResponseEntity<Transaction> transaction) {
+        Map<String, String> map = new HashMap<>();
+        String selfLink = createCompanyAccountSelfLink(entity, transaction);
+        map.put(LinkType.SELF.getLink(), selfLink);
+        entity.getData().setLinks(map);
+    }
+
+    private String createCompanyAccountSelfLink(CompanyAccountEntity entity, ResponseEntity<Transaction> transaction) {
+        return getTransactionSelfLink(transaction) + "/company-accounts/" + entity.getId() + "";
+    }
+
+    private String getTransactionSelfLink(ResponseEntity<Transaction> transaction) {
+        return transaction.getBody().getLinks().get(LinkType.SELF.getLink());
+    }
+
     @Override
     public void addLinks(CompanyAccount rest) {
 
     }
 
     @Override
     public void addKind(CompanyAccount rest) {
-        rest.setKind(Kind.ACCOUNT.getValue());
+        rest.setKind("company-accounts");
+    }
+
+    public void addEntityID(CompanyAccountEntity entity) {
+        entity.setId(UUID.randomUUID().toString());
+    }
+
+    private static HttpServletRequest getRequestFromContext() {
+        return ((ServletRequestAttributes)RequestContextHolder.getRequestAttributes()).getRequest();
+    }
+
+    private static ResponseEntity<Transaction> getTransactionFromSession(HttpServletRequest request) {
+        return (ResponseEntity) request.getSession().getAttribute("transaction");
```

Comment:
"transaction" is being hard coded in multiple places in this PR and https://github.com/companieshouse/company-accounts.api.ch.gov.uk/pull/16. It would be better to place it in a separate class to avoid issues in the future if someone changes it.

Reply:
Will be fixed by Steve in a different PR